docs/userguide/src/portingguide/concerns/weakref.md

```diff
@@ -0,0 +1,477 @@
+2.  In `Scanning::process_weak_refs`, we iterate through all ephemerons in the queue.  If the key of
+    an ephemeron has been reached, but its value has not yet been reached, then retain its value,
+    and remove the ephemeron from the queue.  Otherwise, keep the object in the queue.
+3.  If any value is retained, return `true` from `Scanning::process_weak_refs` so that it will be
+    called again after the transitive closure from retained values are computed.  Then go back to
+    Step 2.
+4.  If no value is retained, the algorithm completes.  The queue contains reachable ephemerons that
+    have unreachable keys.
+
+This algorithm can be modified if we have a list of all ephemerons before GC starts.  We no longer
+need to maintain the queue.
+
+-   In Step 1, we don't need to enqueue ephemerons.
+-   In Step 2, we iterate through all ephemerons.  We retain the value if both the ephemeron itself
+    and the key have been reached, and the value has not been reached, yet.  We don't need to remove
+    any ephemeron from the list.
+-   When the algorithm completes, we can identify both reachable and unreachable ephemerons that
+    have unreachable keys.  But we need to remove unreachable (dead) ephemerons from the list
+    because they will be recycled in the `Release` stage.
+
+And we can go through ephemerons with unreachable keys and do necessary clean-up operations, either
+immediately or postponed to mutator time.
+
+
+## Optimizations
+
+### Generational GC
+
+MMTk provides generational GC plans.  Currently, there are `GenCopy`, `GenImmix` and `StickyImmix`.
+In a minor GC, a generational plan only consider *young objects* (i.e. objects allocated since the
+last GC) as candidates of garbage, and will assume all *old objects* (i.e. objects survived the last
+GC) are live.
+
+The VM binding can query if the current GC is a nursery GC by calling
+
+```rust
+let is_nursery_gc = mmtk.get_plan().generational().is_some_and(|gen|
+    gen.is_current_gc_nursery());
+```
+
+The VM binding can make use of this information when processing finalizers and weak references.  In
+a minor GC,
+
+-   The VM binding only needs to visit **finalizable objects allocated since the last GC**.  Other
+    finalizable objects must be old and will not be considered dead.
+-   The VM binding only needs to visit **weak reference slots written since the last GC**.  Other
+    slots must be pointing to old objects (if not `null`).  For weak hash tables, if existing
+    entries are immutable, it is sufficient to only visit newly added entries.
+
+Implementation-wise, the VM binding can split the lists or hash tables into two parts: one for old
+entries and another for young entries.
+
+### Copying versus non-copying GC
+
+MMTk provides both copying and non-copying GC plans.  `MarkSweep` never moves any objects.
+`MarkCompact`, `SemiSpace` always moves all objects.  Immix-based plans sometimes do non-copying GC,
```

Comment:
Ehhh. That's not true since LOS/non-moving objects are not moved. Should say normal objects there maybe?

Reply:
Sure.  I'll make it clear objects in the LOS, ImmortalSpace, etc. don't move.